dspace-uclouvain/src/main/java/org/dspace/uclouvain/consumer/CataretroFacultyMetadataConsumer.java

```diff
@@ -0,0 +1,159 @@
+    public void end(Context context) throws Exception {}
+
+    @Override
+    public void finish(Context context) throws Exception {}
+
+    /**
+     * Check if an event should be proceeded by this consumer.
+     * 
+     * @param context: The current DSpace context.
+     * @param event: The event to evaluate.
+     * @return True if the event is relevant for this consumer, False otherwise
+     */
+    private Boolean canBeProcessed(Context context, Event event) throws SQLException {
+        if (event.getSubjectType() != Constants.ITEM) {
+            log.warn("CataretroMetadataConsumer should not have been given this kind of subject in an event, skipping: " + event);
+            return false;
+        }
+        Item item = (Item)event.getSubject(context);
+        return item != null && isRelevantCollection(context, item) && isRelevantMetadataModified(event.getDetail());
+    }
+    /** Check if one modified metadata match faculty code metadata field */
+
+    private Boolean isRelevantMetadataModified(String modifiedMetadataFields) {
+        // If the modified fields list is null or empty, it could be because we delete the last "faculty code"
+        // In this case, we need to execute this consumer to delete old faculty names derived from previously
+        // encoded faculty codes.
+        if (modifiedMetadataFields == null || modifiedMetadataFields.trim().isEmpty())
+            return true;
+        // Otherwise, check that any of the modified fields is a faculty code.
+        return Arrays.stream(modifiedMetadataFields.split(","))
+                .map(String::trim)
+                .map(m -> m.replaceAll("_", "."))
+                .anyMatch(x -> x.equals(this.facultyCodeFieldname));
+    }
+    /** Check if the modified item is a member of a "Cataretro" collection */
+
+    private Boolean isRelevantCollection(Context context, Item item) {
+        return getItemCollection(context, item)
+                .stream()
+                .anyMatch(c -> {
+                    List<MetadataValue> metadata = collectionService.getMetadataByMetadataString(c, "dcterms.provenance");
+                    return metadata.stream().map(MetadataValue::getValue).anyMatch(v -> v.equals("cataretro"));
+                });
+    }
+
+    /**
+     * Get collections related to an item.
+     * Depending on item life-cycle, the item collection should be retrieved differently.
+     *
+     * @param context: The DSpace context
+     * @param item: the DSpace Item to analyze.
+     * @return the list of collection to which the item belongs.
+     */
+    private List<Collection> getItemCollection(Context context, Item item) {
+        try {
+            WorkspaceItem wsItem = workspaceItemService.findByItem(context, item);
+            if (wsItem != null) {
+                return Collections.singletonList(wsItem.getCollection());
+            }
+            XmlWorkflowItem wfItem = (XmlWorkflowItem) workflowItemService.findByItem(context, item);
```

Comment:
You could use `XmlWorkflowItemService` instead and you will not have to cast the result.